lib/rdoc/base_plugin.rb

```diff
@@ -0,0 +1,20 @@
+module RDoc
+  class BasePlugin
+    # Register a literner for the given event
+
+    def self.listens_to(event_name, &block)
+      rdoc.event_registry.register(event_name, block)
+    end
+
+    # Activate the plugin with the given RDoc instance
+    # Without calling this, plugins won't work
+
+    def self.activate_with(rdoc = ::RDoc::RDoc.current)
+      @@rdoc = rdoc
```

Comment:
Using class variable for this means any later plugin calls `activate_with` with its own `RDoc` will apply that to all other plugins. I don't think this is what we want?


```rb
class Base
  @@name = "Base"

  def name
    @@name
  end

  def self.update_name(name)
    @@name = name
  end
end

class Foo < Base
end

class Bar < Base
end

Foo.update_name("Foo")
f = Foo.new

Bar.update_name("Bar")
b = Bar.new

puts f.name # => "Bar"
puts b.name # => "Bar"

```

Reply:
I thought it's safe to use a class variable here since we don't have to update `rdoc` object.
However, some information such as a name of the plugin, should be stored as a class instance variables.

Reply:
- If we don't expect plugins to call this method, then I don't think it should be in the base class as it is.
- It doesn't look like the yard plugin needs this attribute yet? If that's the case, let's not introduce this until we have a real use case to help us define the API better.
- If we want to pass RDoc's states down to plugins, let's avoid passing the `RDoc` object unless we have a solid use case for it. There's a lot of unnecessary coupling between RDoc's major components and I have been untangling for a while ([example](https://github.com/ruby/rdoc/pull/1308)). Exposing `RDoc` directly will potentially make it harder for such work as all of its attributes & methods will be considered public APIs.

Reply:
You are right, I didn't notice that it's not used at all now. I used it at first, but `store` has enough information for the plugin so I removed it.
Then we can remove this class variable. And I agree, we should avoid `RDoc::RDoc` object (thank you for untangling it!)

Reply:
Yeah I'd expect we expose `Options` and `Store` objects first if we need to share states with plugins.

Reply:
https://github.com/ruby/rdoc/pull/1321/files#diff-2458b7fcdf31ccd44fa0681ba7e381360b3e4623c9f5409358222e4826c3c60dR6

Sorry it turned out that we do use `rdoc` object here because we need some global data store for event handling.
I still agree that we should avoid `RDoc::RDoc` instance, but then where should we put these event registration logic? Maybe we can introduce something like `PluginManager` and pass it to plugins?